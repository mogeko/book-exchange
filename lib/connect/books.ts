import useSWR, { type SWRResponse } from "swr";
import handleQueryParam, { type QueryParamType } from "@/lib/utils/queryTools";

function useBooks(queryParam: QueryParamType<keyof BookTypes> = {}) {
  const query = queryParam
    ? `/books?${handleQueryParam(queryParam)}`
    : "/books";
  const { data, error }: SWRResponse<BookTypes[]> = useSWR(query);
  return {
    books: data,
    isLoading: !error && !data,
    isError: error,
  };
}

export function useBook(
  id: string | number,
  queryParam: QueryParamType<keyof BookTypes> = {}
) {
  const query = queryParam
    ? `/books/${id}?${handleQueryParam(queryParam)}`
    : `/books/${id}`;
  const { data, error }: SWRResponse<BookTypes> = useSWR(query);
  return {
    book: data,
    isLoading: !error && !data,
    isError: error,
  };
}

export interface BookTypes {
  author: string;
  cover: string;
  description: string;
  id: number;
  published: string;
  publisher: string;
  title: string;
}

export default useBooks;
