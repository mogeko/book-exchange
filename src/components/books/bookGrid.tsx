import Card from "@/components/base/card";
import Pagination from "@/components/pagination";
import useBooks from "@/lib/hooks/useBooks";
import { useState } from "react";

const BookGrid: React.FC<DataProps> = ({ maxPages = 1, page, ...other }) => {
  const [index, setIndex] = useState(0);
  const query = { page: page ?? 1 + index, ...other };
  const { data, isLoading } = useBooks(query);
  // Cache the next page of books
  useBooks({ page: query.page + (index < maxPages - 1 ? 1 : 0), ...other });

  return (
    <div className="flex flex-col">
      <div className="grid grid-cols-2 sm:grid-cols-3 md:grid-cols-5 gap-6 my-8 w-full">
        {isLoading
          ? Array.from({ length: 10 }, (_, i) => <Card.Skeleton key={i} />)
          : data?.map((book, i) => <Card key={i} {...book} />)}
      </div>
      <Pagination length={maxPages} index={index} setIndex={setIndex} />
    </div>
  );
};

type DataProps = {
  maxPages?: number;
} & Parameters<typeof useBooks>[0];

export default BookGrid;
