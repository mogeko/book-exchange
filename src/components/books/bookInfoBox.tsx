import type { BookType } from "@/lib/hooks/useBooks";
import Skeleton from "@/components/base/skeleton";
import Image from "next/image";
import dayjs from "dayjs";

const BookInfo: React.FC<BookInfoProps> = ({ title, cover, mate }) => (
  <div className="my-4 flex w-full gap-3">
    <figure
      className="flex items-center justify-center"
      aria-label="Cover Image"
    >
      <div className="relative w-32 h-48">
        <Image src={cover!} layout="fill" alt={title} />
      </div>
    </figure>
    <div className="flex flex-col prose">
      {Object.entries(mate ?? {}).map(([itemName, value], i) => {
        if (itemName === "paperback" || itemName === "hardcover") {
          return (
            <div key={i} className="inline-flex gap-1 text-base">
              <span className="capitalize">{itemName}:</span>
              <span>{value} pages</span>
            </div>
          );
        }
        if (itemName === "publication_date") {
          return (
            <div key={i} className="inline-flex gap-1 text-base">
              <span className="capitalize">Publication Date:</span>
              <span>{dayjs(value).format("YYYY-MM-DD")}</span>
            </div>
          );
        }
        if (itemName === "isbn") {
          return (
            <div key={i} className="inline-flex gap-1 text-base">
              <span className="uppercase">{itemName}:</span>
              <span>{value}</span>
            </div>
          );
        }
        return (
          <div key={i} className="inline-flex gap-1 text-base">
            <span className="capitalize">{itemName}:</span>
            <span>{value}</span>
          </div>
        );
      })}
    </div>
  </div>
);

export const BookInfoSkeleton: React.FC = () => (
  <div className="my-4 flex w-full gap-3">
    <Skeleton.Square className="relative w-32 aspect-2/3" />
    <div className="flex flex-col gap-2  w-full">
      <Skeleton.Line className="w-2/5 h-4" count={3} />
    </div>
  </div>
);

type BookInfoProps = Partial<Pick<BookType, "mate" | "cover" | "title">>;

export default BookInfo;
