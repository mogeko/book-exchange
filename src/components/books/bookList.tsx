import { LongCard } from "@/components/base/card";
import Alert from "@/components/base/alert";
import { useBooksInfinite } from "@/lib/hooks/useBooks";
import useOnScreen from "@/lib/hooks/useOnScreen";
import { useEffect, useRef } from "react";

const MAX_PAGES = 10;

const BookList: React.FC<ParamProps> = (query) => {
  const { data, size, setSize, ...state } = useBooksInfinite(query);
  const bottomRef = useRef<HTMLDivElement>(null);
  const isVisiable = useOnScreen(bottomRef);

  useEffect(() => {
    if (isVisiable && !state.isValidating && size < MAX_PAGES) {
      setSize(size + 1);
    }
  }, [isVisiable, state, size, setSize]);

  return (
    <>
      <div className="flex flex-col">
        {state.isError && <Alert.Error message="Network Error!" />}
        {state.isLoading ? (
          Array.from({ length: 10 }, (_, i) => <LongCard.Skeleton key={i} />)
        ) : (
          <div>
            {data?.map((books, i) => {
              return books.map((book, j) => {
                return <LongCard key={`${i}-${j}`} {...book} />;
              });
            })}
          </div>
        )}
      </div>
      <div ref={bottomRef} />
    </>
  );
};

type ParamProps = Parameters<typeof useBooksInfinite>[0];

export default BookList;
