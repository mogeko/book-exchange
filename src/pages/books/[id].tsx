import Box from "@/layouts/boxes";
import { DefaultLayout } from "@/layouts/layout";
import { useBook, type BookType } from "@/lib/hooks/useBooks";
import { NextPage } from "next";
import Image from "next/image";
import { useRouter } from "next/router";

const BookPage: NextPage = () => {
  return (
    <DefaultLayout>
      <BookView />
    </DefaultLayout>
  );
};

const BookView: React.FC = () => {
  const { query } = useRouter();
  const { book, isError, isLoading } = useBook(query.id as string);

  if (isError) return <p>Error!</p>;
  if (isLoading) return <p>Loading...</p>;
  return (
    <Box>
      <Box.Header>{book?.title}</Box.Header>
      <div className="my-4 flex w-full gap-3">
        <figure className="flex items-center justify-center">
          <div className="relative w-32 h-48">
            <Image src={book?.cover!} layout="fill" alt={book?.title} />
          </div>
        </figure>
        <div className="flex flex-col">
          {Object.entries(book?.mate ?? {}).map(([itemName, value], i) => {
            if (itemName === "paperback" || itemName === "hardcover") {
              return (
                <div key={i} className="inline-flex gap-1 text-base">
                  <span className="capitalize">{itemName}:</span>
                  <span>{value} pages</span>
                </div>
              );
            }
            if (itemName === "publication_date") {
              return (
                <div key={i} className="inline-flex gap-1 text-base">
                  <span className="capitalize">{itemName}:</span>
                  <span>{new Date(value).toLocaleDateString()}</span>
                </div>
              );
            }
            if (itemName === "isbn") {
              return (
                <div key={i} className="inline-flex gap-1 text-base">
                  <span className="text-sm uppercase">{itemName}:</span>
                  <span>{value}</span>
                </div>
              );
            }
            return (
              <div key={i} className="inline-flex gap-1 text-base">
                <span className="capitalize">{itemName}:</span>
                <span>{value}</span>
              </div>
            );
          })}
        </div>
      </div>
    </Box>
  );
};

export default BookPage;
