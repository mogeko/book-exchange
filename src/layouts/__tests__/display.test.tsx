import { fireEvent, render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import * as books from "@/lib/connect/books";
import Display from "@/layouts/display";

const exampleBook = {
  title: "This is an example book",
  cover: "https://picsum.photos/seed/35613/1280/1920/",
  description: "This is an example book",
  published: "2020-1-1T23:59:59Z",
  publisher: "This is an example publisher",
  author: "No One",
  id: 1,
};

describe("Home", () => {
  afterEach(() => {
    jest.resetAllMocks();
  });

  it("renders a Display", () => {
    const exampleBooks = Array.from({ length: 20 }).map((_) => exampleBook);
    const res = { books: exampleBooks, isError: false, isLoading: false };
    jest.spyOn(books, "default").mockImplementation(() => res);
    const { container } = render(
      <Display title="Example Title" queryParam={{ limit: 20 }} />
    );

    expect(books.default).toBeCalledWith({ limit: 20 });
    expect(
      screen.getByRole("heading", {
        name: /Example Title/i,
      })
    ).toBeInTheDocument();
    expect(container.querySelector(".btn-group")).toBeInTheDocument();
  });

  it("try cleck pages buttons", () => {
    const exampleBooks = Array.from({ length: 20 }).map((_) => exampleBook);
    const res = { books: exampleBooks, isError: false, isLoading: false };
    jest.spyOn(books, "default").mockImplementation(() => res);
    const { container } = render(
      <Display title="Example Title" queryParam={{ limit: 20 }} />
    );

    const buttonGroup = container.querySelector(".btn-group");
    expect(buttonGroup).toBeInTheDocument();
    const buttons = buttonGroup?.querySelectorAll("button");
    expect(buttons).toHaveLength(4);
    expect(buttons?.[1]).toHaveClass("btn-active");
    fireEvent.click(buttons?.[2] as HTMLButtonElement);
    expect(buttons?.[2]).toHaveClass("btn-active");
  });

  it("try cleck last/next buttons", () => {
    const exampleBooks = Array.from({ length: 20 }).map((_) => exampleBook);
    const res = { books: exampleBooks, isError: false, isLoading: false };
    jest.spyOn(books, "default").mockImplementation(() => res);
    const { container } = render(
      <Display title="Example Title" queryParam={{ limit: 20 }} />
    );

    const buttons = container.querySelectorAll(".btn-group button");
    expect(buttons?.[1]).toHaveClass("btn-active");
    fireEvent.click(buttons?.[0] as HTMLButtonElement);
    expect(buttons?.[1]).toHaveClass("btn-active");
    fireEvent.click(buttons?.[3] as HTMLButtonElement);
    expect(buttons?.[2]).toHaveClass("btn-active");
  });

  it("snapshot a Display", () => {
    const exampleBooks = Array.from({ length: 20 }).map((_) => exampleBook);
    const res = { books: exampleBooks, isError: false, isLoading: false };
    jest.spyOn(books, "default").mockImplementation(() => res);
    const { container } = render(
      <Display title="Example Title" queryParam={{ limit: 20 }} />
    );

    expect(container).toMatchSnapshot();
  });

  it("renders a Display when limit < 10", () => {
    const exampleBooks = Array.from({ length: 20 }).map((_) => exampleBook);
    const res = { books: exampleBooks, isError: false, isLoading: false };
    jest.spyOn(books, "default").mockImplementation(() => res);
    const { container } = render(
      <Display title="Example Title" queryParam={{ limit: 9 }} />
    );

    expect(books.default).toBeCalledWith({ limit: 9 });
    expect(
      screen.getByRole("heading", {
        name: /Example Title/i,
      })
    ).toBeInTheDocument();
    expect(container.querySelector(".btn-group")).not.toBeInTheDocument();
  });

  it("snpashot a Display when limit < 10", () => {
    const exampleBooks = Array.from({ length: 20 }).map((_) => exampleBook);
    const res = { books: exampleBooks, isError: false, isLoading: false };
    jest.spyOn(books, "default").mockImplementation(() => res);
    const { container } = render(
      <Display title="Example Title" queryParam={{ limit: 9 }} />
    );

    expect(container).toMatchSnapshot();
  });

  it("renders a home page then error", () => {
    const res = { books: [], isError: true, isLoading: false };
    jest.spyOn(books, "default").mockImplementation(() => res);
    render(<Display title="Example Title" queryParam={{ limit: 20 }} />);

    expect(books.default).toBeCalledWith({ limit: 20 });
    expect(screen.getByText("Network Error!")).toBeInTheDocument();
  });

  it("snapshot a home page then error", () => {
    const res = { books: [], isError: true, isLoading: false };
    jest.spyOn(books, "default").mockImplementation(() => res);
    const { container } = render(
      <Display title="Example Title" queryParam={{ limit: 20 }} />
    );

    expect(container).toMatchSnapshot();
  });

  it("renders a home page then loading", () => {
    const res = { books: [], isError: false, isLoading: true };
    jest.spyOn(books, "default").mockImplementation(() => res);
    const { container } = render(
      <Display title="Example Title" queryParam={{ limit: 20 }} />
    );

    expect(books.default).toBeCalledWith({ limit: 20 });
    expect(container.querySelector("div.animate-pulse")).toBeInTheDocument();
  });

  it("snapshot a home page then loading", () => {
    const res = { books: [], isError: false, isLoading: true };
    jest.spyOn(books, "default").mockImplementation(() => res);
    const { container } = render(
      <Display title="Example Title" queryParam={{ limit: 20 }} />
    );

    expect(container).toMatchSnapshot();
  });
});
