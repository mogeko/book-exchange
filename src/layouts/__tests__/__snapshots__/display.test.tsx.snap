// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home snapshot a Display GridContext 1`] = `
<div>
  <div
    class="flex flex-wrap gap-6 my-8 w-full"
  >
    <div
      class="relative flex flex-col w-28"
    >
      <figure
        class="flex justify-center items-center"
      >
        <a
          class="relative w-full h-[calc(7rem/2*3)]"
          href="/books/1"
        >
          <span
            style="box-sizing: border-box; display: block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px;"
          >
            <img
              alt="This is an example book"
              data-nimg="fill"
              decoding="async"
              src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
              style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
            />
            <noscript />
          </span>
        </a>
      </figure>
      <div
        class="flex flex-col text-sm"
      >
        <a
          class="link no-underline text-primary hover:bg-primary hover:text-primary-content focus:bg-primary-focus focus:text-primary-content"
          href="/books/1"
        >
          <h2
            class="truncate"
          >
            This is an example book
          </h2>
        </a>
        <p
          class="truncate"
        >
          No One
        </p>
      </div>
    </div>
    <div
      class="relative flex flex-col w-28"
    >
      <figure
        class="flex justify-center items-center"
      >
        <a
          class="relative w-full h-[calc(7rem/2*3)]"
          href="/books/1"
        >
          <span
            style="box-sizing: border-box; display: block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px;"
          >
            <img
              alt="This is an example book"
              data-nimg="fill"
              decoding="async"
              src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
              style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
            />
            <noscript />
          </span>
        </a>
      </figure>
      <div
        class="flex flex-col text-sm"
      >
        <a
          class="link no-underline text-primary hover:bg-primary hover:text-primary-content focus:bg-primary-focus focus:text-primary-content"
          href="/books/1"
        >
          <h2
            class="truncate"
          >
            This is an example book
          </h2>
        </a>
        <p
          class="truncate"
        >
          No One
        </p>
      </div>
    </div>
    <div
      class="relative flex flex-col w-28"
    >
      <figure
        class="flex justify-center items-center"
      >
        <a
          class="relative w-full h-[calc(7rem/2*3)]"
          href="/books/1"
        >
          <span
            style="box-sizing: border-box; display: block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px;"
          >
            <img
              alt="This is an example book"
              data-nimg="fill"
              decoding="async"
              src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
              style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
            />
            <noscript />
          </span>
        </a>
      </figure>
      <div
        class="flex flex-col text-sm"
      >
        <a
          class="link no-underline text-primary hover:bg-primary hover:text-primary-content focus:bg-primary-focus focus:text-primary-content"
          href="/books/1"
        >
          <h2
            class="truncate"
          >
            This is an example book
          </h2>
        </a>
        <p
          class="truncate"
        >
          No One
        </p>
      </div>
    </div>
    <div
      class="relative flex flex-col w-28"
    >
      <figure
        class="flex justify-center items-center"
      >
        <a
          class="relative w-full h-[calc(7rem/2*3)]"
          href="/books/1"
        >
          <span
            style="box-sizing: border-box; display: block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px;"
          >
            <img
              alt="This is an example book"
              data-nimg="fill"
              decoding="async"
              src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
              style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
            />
            <noscript />
          </span>
        </a>
      </figure>
      <div
        class="flex flex-col text-sm"
      >
        <a
          class="link no-underline text-primary hover:bg-primary hover:text-primary-content focus:bg-primary-focus focus:text-primary-content"
          href="/books/1"
        >
          <h2
            class="truncate"
          >
            This is an example book
          </h2>
        </a>
        <p
          class="truncate"
        >
          No One
        </p>
      </div>
    </div>
    <div
      class="relative flex flex-col w-28"
    >
      <figure
        class="flex justify-center items-center"
      >
        <a
          class="relative w-full h-[calc(7rem/2*3)]"
          href="/books/1"
        >
          <span
            style="box-sizing: border-box; display: block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px;"
          >
            <img
              alt="This is an example book"
              data-nimg="fill"
              decoding="async"
              src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
              style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
            />
            <noscript />
          </span>
        </a>
      </figure>
      <div
        class="flex flex-col text-sm"
      >
        <a
          class="link no-underline text-primary hover:bg-primary hover:text-primary-content focus:bg-primary-focus focus:text-primary-content"
          href="/books/1"
        >
          <h2
            class="truncate"
          >
            This is an example book
          </h2>
        </a>
        <p
          class="truncate"
        >
          No One
        </p>
      </div>
    </div>
    <div
      class="relative flex flex-col w-28"
    >
      <figure
        class="flex justify-center items-center"
      >
        <a
          class="relative w-full h-[calc(7rem/2*3)]"
          href="/books/1"
        >
          <span
            style="box-sizing: border-box; display: block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px;"
          >
            <img
              alt="This is an example book"
              data-nimg="fill"
              decoding="async"
              src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
              style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
            />
            <noscript />
          </span>
        </a>
      </figure>
      <div
        class="flex flex-col text-sm"
      >
        <a
          class="link no-underline text-primary hover:bg-primary hover:text-primary-content focus:bg-primary-focus focus:text-primary-content"
          href="/books/1"
        >
          <h2
            class="truncate"
          >
            This is an example book
          </h2>
        </a>
        <p
          class="truncate"
        >
          No One
        </p>
      </div>
    </div>
    <div
      class="relative flex flex-col w-28"
    >
      <figure
        class="flex justify-center items-center"
      >
        <a
          class="relative w-full h-[calc(7rem/2*3)]"
          href="/books/1"
        >
          <span
            style="box-sizing: border-box; display: block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px;"
          >
            <img
              alt="This is an example book"
              data-nimg="fill"
              decoding="async"
              src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
              style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
            />
            <noscript />
          </span>
        </a>
      </figure>
      <div
        class="flex flex-col text-sm"
      >
        <a
          class="link no-underline text-primary hover:bg-primary hover:text-primary-content focus:bg-primary-focus focus:text-primary-content"
          href="/books/1"
        >
          <h2
            class="truncate"
          >
            This is an example book
          </h2>
        </a>
        <p
          class="truncate"
        >
          No One
        </p>
      </div>
    </div>
    <div
      class="relative flex flex-col w-28"
    >
      <figure
        class="flex justify-center items-center"
      >
        <a
          class="relative w-full h-[calc(7rem/2*3)]"
          href="/books/1"
        >
          <span
            style="box-sizing: border-box; display: block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px;"
          >
            <img
              alt="This is an example book"
              data-nimg="fill"
              decoding="async"
              src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
              style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
            />
            <noscript />
          </span>
        </a>
      </figure>
      <div
        class="flex flex-col text-sm"
      >
        <a
          class="link no-underline text-primary hover:bg-primary hover:text-primary-content focus:bg-primary-focus focus:text-primary-content"
          href="/books/1"
        >
          <h2
            class="truncate"
          >
            This is an example book
          </h2>
        </a>
        <p
          class="truncate"
        >
          No One
        </p>
      </div>
    </div>
    <div
      class="relative flex flex-col w-28"
    >
      <figure
        class="flex justify-center items-center"
      >
        <a
          class="relative w-full h-[calc(7rem/2*3)]"
          href="/books/1"
        >
          <span
            style="box-sizing: border-box; display: block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px;"
          >
            <img
              alt="This is an example book"
              data-nimg="fill"
              decoding="async"
              src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
              style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
            />
            <noscript />
          </span>
        </a>
      </figure>
      <div
        class="flex flex-col text-sm"
      >
        <a
          class="link no-underline text-primary hover:bg-primary hover:text-primary-content focus:bg-primary-focus focus:text-primary-content"
          href="/books/1"
        >
          <h2
            class="truncate"
          >
            This is an example book
          </h2>
        </a>
        <p
          class="truncate"
        >
          No One
        </p>
      </div>
    </div>
    <div
      class="relative flex flex-col w-28"
    >
      <figure
        class="flex justify-center items-center"
      >
        <a
          class="relative w-full h-[calc(7rem/2*3)]"
          href="/books/1"
        >
          <span
            style="box-sizing: border-box; display: block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px;"
          >
            <img
              alt="This is an example book"
              data-nimg="fill"
              decoding="async"
              src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
              style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
            />
            <noscript />
          </span>
        </a>
      </figure>
      <div
        class="flex flex-col text-sm"
      >
        <a
          class="link no-underline text-primary hover:bg-primary hover:text-primary-content focus:bg-primary-focus focus:text-primary-content"
          href="/books/1"
        >
          <h2
            class="truncate"
          >
            This is an example book
          </h2>
        </a>
        <p
          class="truncate"
        >
          No One
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`Home snapshot a Display GridContext then error 1`] = `
<div>
  <div
    class="flex flex-wrap gap-6 my-8 w-full"
  >
    <div
      class="alert alert-error shadow-lg"
    >
      <div>
        <svg
          class="w-6 h-6"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 16 16"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M8.6 1c1.6.1 3.1.9 4.2 2 1.3 1.4 2 3.1 2 5.1 0 1.6-.6 3.1-1.6 4.4-1 1.2-2.4 2.1-4 2.4-1.6.3-3.2.1-4.6-.7-1.4-.8-2.5-2-3.1-3.5C.9 9.2.8 7.5 1.3 6c.5-1.6 1.4-2.9 2.8-3.8C5.4 1.3 7 .9 8.6 1zm.5 12.9c1.3-.3 2.5-1 3.4-2.1.8-1.1 1.3-2.4 1.2-3.8 0-1.6-.6-3.2-1.7-4.3-1-1-2.2-1.6-3.6-1.7-1.3-.1-2.7.2-3.8 1-1.1.8-1.9 1.9-2.3 3.3-.4 1.3-.4 2.7.2 4 .6 1.3 1.5 2.3 2.7 3 1.2.7 2.6.9 3.9.6zM7.9 7.5L10.3 5l.7.7-2.4 2.5 2.4 2.5-.7.7-2.4-2.5-2.4 2.5-.7-.7 2.4-2.5-2.4-2.5.7-.7 2.4 2.5z"
            fill-rule="evenodd"
          />
        </svg>
        <span>
          Network Error!
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`Home snapshot a Display GridContext then loading 1`] = `
<div>
  <div
    class="flex flex-wrap gap-6 my-8 w-full"
  >
    <div
      class="relative flex flex-col mb-4 w-28 animate-pulse"
    >
      <div
        class="mb-2 w-28 h-[calc(7rem/2*3)] bg-slate-700 rounded"
      />
      <div
        class="mb-2 w-24 h-2 bg-slate-700 rounded"
      />
      <div
        class="mb-2 w-20 h-2 bg-slate-700 rounded"
      />
    </div>
    <div
      class="relative flex flex-col mb-4 w-28 animate-pulse"
    >
      <div
        class="mb-2 w-28 h-[calc(7rem/2*3)] bg-slate-700 rounded"
      />
      <div
        class="mb-2 w-24 h-2 bg-slate-700 rounded"
      />
      <div
        class="mb-2 w-20 h-2 bg-slate-700 rounded"
      />
    </div>
    <div
      class="relative flex flex-col mb-4 w-28 animate-pulse"
    >
      <div
        class="mb-2 w-28 h-[calc(7rem/2*3)] bg-slate-700 rounded"
      />
      <div
        class="mb-2 w-24 h-2 bg-slate-700 rounded"
      />
      <div
        class="mb-2 w-20 h-2 bg-slate-700 rounded"
      />
    </div>
    <div
      class="relative flex flex-col mb-4 w-28 animate-pulse"
    >
      <div
        class="mb-2 w-28 h-[calc(7rem/2*3)] bg-slate-700 rounded"
      />
      <div
        class="mb-2 w-24 h-2 bg-slate-700 rounded"
      />
      <div
        class="mb-2 w-20 h-2 bg-slate-700 rounded"
      />
    </div>
    <div
      class="relative flex flex-col mb-4 w-28 animate-pulse"
    >
      <div
        class="mb-2 w-28 h-[calc(7rem/2*3)] bg-slate-700 rounded"
      />
      <div
        class="mb-2 w-24 h-2 bg-slate-700 rounded"
      />
      <div
        class="mb-2 w-20 h-2 bg-slate-700 rounded"
      />
    </div>
    <div
      class="relative flex flex-col mb-4 w-28 animate-pulse"
    >
      <div
        class="mb-2 w-28 h-[calc(7rem/2*3)] bg-slate-700 rounded"
      />
      <div
        class="mb-2 w-24 h-2 bg-slate-700 rounded"
      />
      <div
        class="mb-2 w-20 h-2 bg-slate-700 rounded"
      />
    </div>
    <div
      class="relative flex flex-col mb-4 w-28 animate-pulse"
    >
      <div
        class="mb-2 w-28 h-[calc(7rem/2*3)] bg-slate-700 rounded"
      />
      <div
        class="mb-2 w-24 h-2 bg-slate-700 rounded"
      />
      <div
        class="mb-2 w-20 h-2 bg-slate-700 rounded"
      />
    </div>
    <div
      class="relative flex flex-col mb-4 w-28 animate-pulse"
    >
      <div
        class="mb-2 w-28 h-[calc(7rem/2*3)] bg-slate-700 rounded"
      />
      <div
        class="mb-2 w-24 h-2 bg-slate-700 rounded"
      />
      <div
        class="mb-2 w-20 h-2 bg-slate-700 rounded"
      />
    </div>
    <div
      class="relative flex flex-col mb-4 w-28 animate-pulse"
    >
      <div
        class="mb-2 w-28 h-[calc(7rem/2*3)] bg-slate-700 rounded"
      />
      <div
        class="mb-2 w-24 h-2 bg-slate-700 rounded"
      />
      <div
        class="mb-2 w-20 h-2 bg-slate-700 rounded"
      />
    </div>
    <div
      class="relative flex flex-col mb-4 w-28 animate-pulse"
    >
      <div
        class="mb-2 w-28 h-[calc(7rem/2*3)] bg-slate-700 rounded"
      />
      <div
        class="mb-2 w-24 h-2 bg-slate-700 rounded"
      />
      <div
        class="mb-2 w-20 h-2 bg-slate-700 rounded"
      />
    </div>
  </div>
</div>
`;

exports[`Home snapshot a Display Pagination 1`] = `
<div>
  <div
    class="btn-group"
  >
    <button
      class="btn btn-xs"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 192 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M192 127.338v257.324c0 17.818-21.543 26.741-34.142 14.142L29.196 270.142c-7.81-7.81-7.81-20.474 0-28.284l128.662-128.662c12.599-12.6 34.142-3.676 34.142 14.142z"
        />
      </svg>
    </button>
    <button
      class="btn btn-xs btn-active"
      value="0"
    >
      1
    </button>
    <button
      class="btn btn-xs"
      value="1"
    >
      2
    </button>
    <button
      class="btn btn-xs"
      value="2"
    >
      3
    </button>
    <button
      class="btn btn-xs"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 192 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 384.662V127.338c0-17.818 21.543-26.741 34.142-14.142l128.662 128.662c7.81 7.81 7.81 20.474 0 28.284L34.142 398.804C21.543 411.404 0 402.48 0 384.662z"
        />
      </svg>
    </button>
  </div>
</div>
`;
